### PR TITLE
[DR-2537] Add new datasets "properties" field

### PR DESCRIPTION
**Pull request by samanehsan**
This PR adds a new optional "properties" JSON field to datasets that allows users to store additional metadata about a dataset. This can be set when creating a dataset and/or updated using the dataset patch endpoint. Note that the patch endpoint will do a full replace of the value. When getting dataset details, the properties are not included unless `PROPERTIES` is a part of the `includes` list.

### CODE REVIEW

File: src/main/java/bio/terra/common/DaoUtils.java [hunk 1]
```diff
@@ -113,4 +115,17 @@ public static boolean retryQuery(DataAccessException dataAccessException) {
     return ExceptionUtils.hasCause(dataAccessException, RecoverableDataAccessException.class)
         || ExceptionUtils.hasCause(dataAccessException, TransientDataAccessException.class);
   }
+
+  public static PGobject propertiesToPGobject(ObjectMapper objectMapper, Object properties) {
```
Comment (nmalfroy) [inline src/main/java/bio/terra/common/DaoUtils.java]: That's interesting: I'm actually kinda surprised that you had to do this.  Did writing the data as a vanilla JSON string not work?

File: src/main/java/bio/terra/service/dataset/DatasetDao.java [hunk 2]
```diff
@@ -714,10 +725,17 @@ public DatasetSummary mapRow(ResultSet rs, int rowNum) throws SQLException {
...
+        "UPDATE dataset SET phs_id = COALESCE(:phs_id, phs_id), "
+            + "properties = COALESCE(:properties, properties) WHERE id = :id";
 
     MapSqlParameterSource params =
-        new MapSqlParameterSource().addValue("phs_id", patchRequest.getPhsId()).addValue("id", id);
+        new MapSqlParameterSource()
+            .addValue("phs_id", patchRequest.getPhsId())
+            .addValue(
+                "properties",
+                DaoUtils.propertiesToPGobject(objectMapper, patchRequest.getProperties()))
```
Comment (nmalfroy) [inline src/main/java/bio/terra/service/dataset/DatasetDao.java]: What happens if the patch request is null? Just want to make sure it doesn't do something like have an explicit `'null'` string

**Review by nmalfroy — APPROVED** [review]
This looks great!  I was just wondering about making sure that you need that bit that explicitly sets it to a jsonb type of object.  I thought postgres did that for us on the back end automatically when we send in a string?  If we don't need it, it might make sense to take it out (avoids the risk of NPEs) but I'll leave it to you

File: src/test/java/bio/terra/service/dataset/DatasetDaoTest.java [hunk 3]
```diff
@@ -877,4 +878,47 @@ public void patchDatasetPhsId() throws Exception {
...
+  }
+
+  @Test
+  public void patchDatasetProperties() throws Exception {
+    UUID datasetId = createDataset("dataset-create-test.json");
+    assertThat(
+        "dataset properties is null before patch",
+        datasetDao.retrieve(datasetId).getProperties(),
+        equalTo(null));
+
```
Comment (okotsopoulos) [inline src/test/java/bio/terra/service/dataset/DatasetDaoTest.java:902]: It would be nice to see another test case here: if the existing `properties` value is null and a patch is applied which does not alter `properties`, the `properties` value remains null.

File: src/main/java/bio/terra/common/DaoUtils.java (hunk 1 above)
Comment (pshapiro4broad) [reply src/main/java/bio/terra/common/DaoUtils.java]: I don't know if what catalog is doing is better or not, but we cast to jsonb inline in the SQL, and don't use a custom mapper. 

https://github.com/DataBiosphere/terra-data-catalog/blob/main/service/src/main/java/bio/terra/catalog/service/dataset/DatasetDao.java#L83

Comment (pshapiro4broad) [reply src/main/java/bio/terra/common/DaoUtils.java]: In our case we have `String` which can be cast to `jsonb`, maybe if you have `Object` you have to first render to `String` before you can write it to postgres. That would make sense.

File: src/main/resources/api/data-repository-openapi.yaml [hunk 4]
```diff
@@ -3340,6 +3340,9 @@ components:
           type: boolean
           default: false
           description: denotes whether data files in the dataset are self-hosted or not
+        properties:
+          type: object
+          description: Additional metadata about the dataset.
```
Comment (okotsopoulos) [inline src/main/resources/api/data-repository-openapi.yaml]: As a user, I might find a little more elaboration here helpful.  The "blank check" nature of this field is almost too good to be true!  Otherwise I might assume that there is some validated format or acceptable list of keys that I don't know about.

**Review by okotsopoulos — APPROVED** [review]
Looks good!  A few small requests.  It's great to see additional use out of the patch endpoints!

File: src/main/java/bio/terra/common/DaoUtils.java (hunk 1 above)
Comment (samanehsan) [reply src/main/java/bio/terra/common/DaoUtils.java]: Postgres thought it was an hstore data type when I tried to just use a JSON string. I'll try casting it inline!

File: src/main/java/bio/terra/service/dataset/DatasetDao.java (hunk 2 above)
Comment (samanehsan) [reply src/main/java/bio/terra/service/dataset/DatasetDao.java]: Good catch! It was setting the value to "null" so I fixed that and added some more tests!

File: src/main/java/bio/terra/common/DaoUtils.java (hunk 1 above)
Comment (samanehsan) [reply src/main/java/bio/terra/common/DaoUtils.java]: @pshapiro4broad casting a JSON string worked! Thanks for the suggestion 🙂
